filecoin-proofs/src/api/mod.rs

```diff
@@ -58,80 +53,40 @@ pub use winning_post::*;
 pub use storage_proofs_update::constants::partition_count;
 
 // Ensure that any associated cached data persisted is discarded.
-pub fn clear_cache<Tree: MerkleTreeTrait>(cache_dir: &Path) -> Result<()> {
+pub fn clear_cache(cache_dir: &Path) -> Result<()> {
```

Comment:
Note that `clear_cache`, `clear_layer_data`, and `clear_synthetic_proofs` are exposed API methods, not just internal.  In other words, dropping (or potentially changing) them will cause issues to users of them.

Reply:
Yes, I know. `clear_cache` and `clear_layer_data` did the same thing, hence one was dropped (of course it need to be reflected in proofs API as well as FFI. The way they work is indeed slightly different. Two things that are different:

 - Trees which consist of several files are now deleted by a glob. So in case someone would have a file that is names similarly to a tree file, but is not really part of it, that file would be deleted. Does that really happen?
 - The trees were opened. If I understand it correctly, then it could have happened that if a file was corrupt that it wasn't deleted. Now it's just deleted. I think that's a win.

I would argue that such a massive simplification of the code, and also for people to explain "xyz" will happen justifies that change.

Reply:
The API signatures are different, so I'd prefer we not disturb them for now.  If they weren't exposed, it's an easy internal re-factor.  They didn't do the same thing originally, but ended up doing the same btw.  They may not again at some point later.

I don't think glob delete is a good idea because we don't 'own' the directory we're using, necessarily.  If it was entirely created and managed internally, it could be different.  I see it as our API is given access to a dir in which we persist and use the things we require and leave whatever else may be there up to others.

Reply:
> The API signatures are different, so I'd prefer we not disturb them for now. If they weren't exposed, it's an easy internal re-factor. They didn't do the same thing originally, but ended up doing the same btw. They may not again at some point later.
> 
> I don't think glob delete is a good idea because we don't 'own' the directory we're using, necessarily. If it was entirely created and managed internally, it could be different. I see it as our API is given access to a dir in which we persist and use the things we require and leave whatever else may be there up to others.

So it seems like the difference was needed => we should take the chance to make a change and unify it into a single function all the way up. It it turns out something else is needed in the future, it can be added again.



> The API signatures are different, so I'd prefer we not disturb them for now. If they weren't exposed, it's an easy internal re-factor. They didn't do the same thing originally, but ended up doing the same btw. They may not again at some point later.

 
> I don't think glob delete is a good idea because we don't 'own' the directory we're using, necessarily. If it was entirely created and managed internally, it could be different. I see it as our API is given access to a dir in which we persist and use the things we require and leave whatever else may be there up to others.

I would solve this by communicating in a change long that files with such and such prefix are going to be removed.

The problem if we don't use globs is, that we either need a new API for the "no t_aux" case or we have different behaviour for the "no t_aux" case, which would be painful to maintain.

I totally agree that we should not break things where it isn't needed. But IMHO that's really more of theoretic breakage then practical one. It's not like we delete all files. We delete files with a specific prefix and extension.

Would an option be to discuss this with the Lotus team and see if they would be OK with that change?

Reply:
> The API signatures are different, so I'd prefer we not disturb them for now. If they weren't exposed, it's an easy internal re-factor. They didn't do the same thing originally, but ended up doing the same btw. They may not again at some point later.
> 
> I don't think glob delete is a good idea because we don't 'own' the directory we're using, necessarily. If it was entirely created and managed internally, it could be different. I see it as our API is given access to a dir in which we persist and use the things we require and leave whatever else may be there up to others.

I had another look. The `clear_caches()` function is not used at all. Not in `rust-fil-proofs` and not in `rust-filecoin-proofs-api`. So I'd really be in favour of removing it. let's take the chance and reducing the API surface in cases it isn't even used. I can also open a separate PR if you like (which I'd actually prefer).

`clear_layer_data()` and `clear_cache()` have the same signature and do the same thing. The `clear_layer_data()` is exposed through `rust-filecoin-proofs-api`, so yes, it's a breaking change, but we can easily just call `clear_cache()` directly from `rust-filecoin-proofs-api` or (which I'd prefer and happy to do a PR` remove it there. I really prefer if we don't drag on cruft, which isn't really needed. We should strive for a minimal API surface where possible.

Reply:
Not following the latest here, but `clear_cache` is exposed via FFI as well (saying it's not used in an `api` test doesn't mean it's not used)

Reply:
Looks like I searched accidentally for `clear_caches` (with the "s"). Yes you're right, it's part of the FFI.